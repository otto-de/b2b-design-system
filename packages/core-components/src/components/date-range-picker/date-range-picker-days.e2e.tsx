import type { E2EPage } from '@stencil/core/testing';
import { newE2EPage } from '@stencil/core/testing';

describe('B2B-Date-Range-Picker-Days', () => {
  let page: E2EPage;
  beforeEach(async () => {
    page = await newE2EPage();
  });

  it('should emit the clicked date', async () => {
    await page.setContent(
      `<b2b-date-range-picker-days viewed-date="01.01.2025"></b2b-date-range-picker-days>`,
    );
    await page.waitForChanges();

    const dateSelectedSpy = await page.spyOnEvent('b2b-date-selected');

    const dayComponent = await page.find(
      'b2b-date-range-picker-days >>> .b2b-date-range-picker-day',
    );
    await dayComponent.click();
    page.waitForChanges();

    expect(dateSelectedSpy).toHaveReceivedEvent();
    expect(dateSelectedSpy).toHaveReceivedEventDetail({
      selectedDate: new Date(2025, 0, 1).toISOString(), // Must be a string due to node-to-browser "serialization"
    });
  });
});
