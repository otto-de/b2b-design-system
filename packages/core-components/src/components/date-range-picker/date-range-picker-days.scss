@use 'sass:color';
@use '../../global/b2b-styles';

.b2b-date-range-picker-days {
  display: grid;
  grid-template-columns: repeat(7, 1fr); /* Display days in a grid */
  padding: 0 var(--b2b-size-40) 0 var(--b2b-size-40);

  .b2b-date-range-picker-day {
    width: var(--b2b-size-90);
    height: var(--b2b-size-90);
    border-radius: 5px;
    display: flex;
    align-items: center;
    justify-content: center;
    font-weight: var(--b2b-font-weight-normal);
    color: var(--b2b-color-black-100);
    font-size: var(--b2b-size-35);
    cursor: pointer;

    &:hover {
      z-index: 102;
    }

    &--disabled {
      color: var(--b2b-color-grey-200);
      pointer-events: none;
    }

    &--today {
      border: 1px solid var(--b2b-color-black-50);
      box-sizing: border-box;
    }

    &--selected {
      background-color: var(--b2b-color-black-100);
      color: var(--b2b-color-white-100);
    }

    &--range {
      background-color: var(--b2b-color-grey-200);
      color: var(--b2b-color-black-100);
    }

    &--preview.b2b-date-range-picker-day--range,
    &:hover,
    &:focus {
      background-color: var(--b2b-color-hover-default);
      color: var(--b2b-color-black-100);
    }
  }
}
