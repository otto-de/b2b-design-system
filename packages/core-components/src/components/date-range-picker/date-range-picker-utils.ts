// #region Dateish

/**
 * A `Date`-like value.
 */
export type Dateish = string | Date | DateTuple;

/**
 * The main constituents of the date part of a `Date`.
 */
export type DateTuple = readonly [number, number, number];

/**
 * Splits the given date string into its constituent parts.
 * @param date The date to split.
 * @returns A tuple of the given year, month and day of the given date (Month and days are 1-indexed).
 */
export function splitDate(date: Dateish): DateTuple;
export function splitDate(date: Dateish | undefined): DateTuple | undefined;
export function splitDate(date: Dateish): DateTuple {
  if (date == null) {
    return undefined;
  } else if (typeof date === 'string') {
    if (date.includes('T')) {
      try {
        return splitDate(new Date(date));
      } catch {
        throw new Error(`Invalid date: ${date}`);
      }
    }

    const parts = date.split('.').map(Number);
    if (parts.length !== 3) {
      throw new Error(`Invalid date: ${date}`);
    }
    const [day, month, year] = date.split('.').map(Number);
    return [year, month, day];
  } else if (isDateTuple(date)) {
    return date;
  } else if (date instanceof Date && !Number.isNaN(date.valueOf())) {
    return [date.getFullYear(), date.getMonth() + 1, date.getDate()];
  } else {
    throw new Error(`Invalid date: ${date}`);
  }
}

/**
 * Converts the given date in a number to simplify comparisons.
 * @param date The date to convert.
 * @returns The given date as a number for comparisons .
 */
export function dateAsNumber(date: Dateish): number;
export function dateAsNumber(date: Dateish | undefined): number | undefined;
export function dateAsNumber(date: Dateish): number {
  if (date == null) {
    return undefined;
  } else {
    const [year, month, day] = splitDate(date);
    return year * 10000 + month * 100 + day;
  }
}

/**
 * Checks whether the given value is a `DateTuple`.
 * @param value The value to check.
 * @returns True, if the value is a `DateTuple`. False otherwise.
 */
function isDateTuple(value: unknown): value is DateTuple {
  return Array.isArray(value) && value.length === 3;
}

export function toDate(date: Dateish): Date;
export function toDate(date: Dateish | undefined): Date | undefined;
export function toDate(date: Dateish | undefined): Date | undefined {
  if (date == null) {
    return undefined;
  } else {
    const [year, month, day] = splitDate(date);
    return new Date(year, month - 1, day);
  }
}

export function toDateString(date: Dateish, nullDefault?: string): string;
export function toDateString(
  date: Dateish | undefined,
  nullDefault: string,
): string;
export function toDateString(
  date: Dateish | undefined,
  nullDefault?: string,
): string | undefined;
export function toDateString(date: Dateish, nullDefault?: string): string {
  if (date == null) {
    return nullDefault;
  } else {
    const [year, month, day] = splitDate(date);
    return `${day.toString().padStart(2, '0')}.${month.toString().padStart(2, '0')}.${year}`;
  }
}

/**
 * Checks whether the given dates are exactly equal (down to the ms).
 * @param first The first date to compare.
 * @param second The second date to compare.
 * @returns True, if both values are exactly equal or both nullish. False otherwise.
 */
function isDateEqual(first?: Date, second?: Date): boolean {
  return first?.valueOf() === second?.valueOf();
}

// #endregion
// #region DateRange

/**
 * A readonly tuple of two `Date`s.
 */
export type DateRange = readonly [Date, Date];

/**
 * A readonly tuple of two `Date`-ish values.
 */
export type DatishRange = readonly [Dateish, Dateish];

/**
 * Parses the given input as a `DateRange`.
 * @param range The input to parse.
 * @returns The parsed date range. May be `undefined` if the input is nullish.
 */
export function toDateRange(
  range: string | readonly [Dateish, Dateish],
): DateRange;
export function toDateRange(
  range: undefined | string | readonly [Dateish, Dateish],
): undefined | DateRange;
export function toDateRange(
  range: undefined | string | readonly [Dateish, Dateish],
): undefined | DateRange {
  if (range == null) {
    return undefined;
  } else if (typeof range === 'string') {
    try {
      return toDateRange(JSON.parse(range));
    } catch {
      let parts = range.split('-');
      if (parts.length !== 2) {
        parts = range.split(' - ');
        if (parts.length !== 2) {
          throw new Error(`Invalid date range: ${range}`);
        }
      }
      return toDateRange([parts[0], parts[1]]);
    }
  } else if (Array.isArray(range) && range.length === 2) {
    const [start, end] = range;
    return [toDate(start), toDate(end)];
  } else {
    throw new Error(`Invalid date range: ${range}`);
  }
}

/**
 *Converts the given date range into a string.
 * @param range The value to convert.
 * @param nullDefault The value to use in case the input is nullish.
 * @returns A string representation of the given input.
 */
export function toDateRangeString(
  range: DatishRange,
  nullDefault?: string,
): string;
export function toDateRangeString(
  range: DatishRange | undefined,
  nullDefault: string,
): string;
export function toDateRangeString(
  range: DatishRange | undefined,
  nullDefault?: string,
): string | undefined;
export function toDateRangeString(
  range: DatishRange | undefined,
  nullDefault?: string,
): string | undefined {
  if (range == null) {
    return nullDefault;
  } else {
    return `${toDateString(range[0], '??.??.????')} - ${toDateString(range[1], '??.??.????')}`;
  }
}

/**
 * Checks whether the given date ranges are exactly equal (down to the ms).
 * @param first The first date range to compare.
 * @param second The second date range to compare.
 * @returns True, if both values are exactly equal or both nullish. False otherwise.
 */
export function isDateRangeEqual(
  first: DateRange | undefined,
  second: DateRange | undefined,
): boolean {
  if (first == null) {
    return second == null;
  } else if (second == null) {
    return false;
  }

  return isDateEqual(first[0], second[0]) && isDateEqual(first[1], second[1]);
}

// #endregion
