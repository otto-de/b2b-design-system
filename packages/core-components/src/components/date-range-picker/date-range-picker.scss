@use '../../global/b2b-styles';

:host {
  position: relative;
}

.b2b-date-range-picker {
  position: relative;

  &--backdrop {
    position: fixed;
    top: 0;
    left: 0;
    width: 100%;
    height: 100%;
    z-index: 100;
  }
}

.b2b-date-range-picker-input-wrapper {
  display: flex;
  box-sizing: border-box;
  height: var(--b2b-size-90); /* The height of the wrapper */
  border: var(--b2b-size-1) solid var(--b2b-color-grey-200);
  justify-content: center;
  cursor: text;
  border-radius: var(--b2b-size-7);
  padding: var(--b2b-size-13) var(--b2b-size-25) var(--b2b-size-13);
  background-color: var(--b2b-color-white-100);
  transition:
    outline 0.2s ease-in-out,
    border-color 0.2s ease-in-out;

  &--error {
    background-color: var(--b2b-color-error-50);
    border: 1px solid var(--b2b-color-error-100);
    border-color: var(--b2b-color-error-100);
  }

  b2b-paragraph {
    width: 100%;
  }

  .b2b-icons {
    display: flex;
    width: 48px; /* two icons wide to prevent shifting when the clear logo (disappears) */
    justify-content: end;
  }
}

.b2b-date-range-picker-hint {
  font-size: var(--b2b-size-copy-50);
  line-height: var(--b2b-size-copy-line-height-75);
  display: block;
  margin-top: var(--b2b-size-7);
  color: var(--b2b-color-grey-300);

  &--error {
    color: var(--b2b-color-error-100);
  }
}

.b2b-date-range-picker-body {
  position: absolute;
  display: flex;
  flex-direction: row;
  outline-offset: -1px;
  box-sizing: border-box;
  margin-top: 1px;
  outline: none;
  background-color: var(--b2b-color-white-100);
  z-index: 101;

  &--hidden {
    display: none;
  }

  &--calendar {
    border-right: var(--b2b-size-1) solid var(--b2b-color-black-100);
    border-left: var(--b2b-size-1) solid var(--b2b-color-black-100);
    border-bottom: var(--b2b-size-1) solid var(--b2b-color-black-100);
    padding-bottom: var(--b2b-size-40);
    width: 300px;
  }

  &--presets {
    padding: var(--b2b-size-30) var(--b2b-size-40) var(--b2b-size-40)
      var(--b2b-size-40);
    border: var(--b2b-size-1) solid var(--b2b-color-grey-25);
    display: flex;
    flex-direction: column;
    gap: 4px;

    b2b-paragraph {
      cursor: pointer;
      width: max-content;
      border-radius: var(--b2b-size-7);
      padding: var(--b2b-size-7) var(--b2b-size-13);

      &:hover {
        background-color: var(--b2b-color-hover-default);

        --b2b-color-copy-default: var(--b2b-color-black-100);

        color: var(
          --b2b-color-copy-default
        ); /* required to prevent discarding of the variable overwrite */
      }
    }

    &--selected {
      background-color: var(--b2b-color-black-100);

      --b2b-color-copy-default: var(--b2b-color-white-100);

      color: var(
        --b2b-color-copy-default
      ); /* required to prevent discarding of the variable overwrite */
    }
  }
}
