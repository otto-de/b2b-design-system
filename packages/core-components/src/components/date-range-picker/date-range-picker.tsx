import type { ComponentInterface, EventEmitter } from '@stencil/core';
import {
  Component,
  Event,
  h,
  Host,
  Listen,
  Prop,
  State,
  Watch,
} from '@stencil/core';
import type {
  DateClear,
  DatePickerEventDetail,
  DatePickerViewChangedEventDetail,
  DateRangePickerEventDetail,
  MonthSelectedEventDetail,
  YearSelectedEventDetail,
} from '../../utils/interfaces/form.interface';
import { DatePickerView } from '../date-picker/date-picker.types';
import type {
  Dateish,
  DateRange,
  DateTuple,
  DatishRange,
} from './date-range-picker-utils';
import {
  dateAsNumber,
  isDateRangeEqual,
  splitDate,
  toDate,
  toDateRange,
  toDateRangeString,
} from './date-range-picker-utils';

@Component({
  tag: 'b2b-date-range-picker',
  styleUrl: 'date-range-picker.scss',
  shadow: true,
})
export class B2bDateRangePicker implements ComponentInterface {
  // #region Properties

  /**
   * Label for the date range picker component.
   *
   * @default 'Zeitraum auswählen'
   */
  @Prop() label: string = 'Zeitraum auswählen';

  /**
   * Adds an asterisk at the end of the label to signify that the field is required.
   *
   * @default false
   */
  @Prop() required: boolean = false;

  /**
   * The start and the end date of the selected date range.
   * Format: `TT.MM.YYYY`
   */
  @Prop() dateRange?: string | DatishRange;

  /**
   * Function used to determine, whether a date should be disabled e.g. because it is in the past.
   * This is only used for the start and end dates, it is valid to have a disabled date in the range of start and end date.
   *
   * @param date The date to check. Format: `TT.MM.YYYY`
   * @returns True, if the date should be disabled. False otherwise.
   *
   * @default (date) => false
   */
  @Prop() disableDates: (date: Date) => boolean = () => false;

  /**
   * A list of presets, that the user can click on to simplify time range selection.
   *
   * @default {}
   */
  @Prop() presets: Record<string, DateRange> = {};

  /**
   * Whether the input is currently invalid. If true, the input is rendered with error styles.
   *
   * @default false
   */
  @Prop() invalid: boolean = false;

  /**
   * The hint text that appears underneath the input field.
   */
  @Prop() hint?: string;

  /**
   * The error message that is shown if the input is invalid.
   */
  @Prop() error?: string;

  /**
   * The language for month and the weekdays will be decided based on the given input.
   * One of `'de'` (German) or `'English'`.
   *
   * @default 'de'
   */
  @Prop() language: 'de' | 'en' = 'de';

  // #endregion
  // #region Events

  /**
   * Emits the selected date range as tuple type.
   */
  @Event({ eventName: 'b2b-selected' })
  b2bSelected: EventEmitter<DateRangePickerEventDetail>;

  /**
   * Emits when the user clicks the clear button.
   */
  @Event({ eventName: 'b2b-clear' })
  b2bClear: EventEmitter<DateClear>;

  // #endregion
  // #region State

  @State() private open: boolean = false;
  @State() private tempDate: DateTuple | undefined = undefined;
  @State() private internalDateRange?: DateRange;
  @State() private highlightDateRange?: DateRange;
  @State() private viewMode: DatePickerView = DatePickerView.Days;
  @State() private viewedYear: number;
  @State() private viewedMonth: number; // 1-based
  @State() private viewedDay: number; // 1-based

  @Watch('dateRange')
  updateDateRange(value: DateRange | undefined): void {
    this.internalDateRange = toDateRange(value);
    this.resetViewedDate();
  }

  // #endregion
  // #region Logic

  @Listen('b2b-date-picker-view-changed')
  handleDatePickerViewChanged(
    event: CustomEvent<DatePickerViewChangedEventDetail>,
  ): void {
    this.viewMode = event.detail.value;
  }

  @Listen('b2b-date-picker-previous-month')
  previousMonth(): void {
    if (this.viewedMonth === 0) {
      this.viewedMonth = 11;
      this.viewedYear -= 1;
    } else {
      this.viewedMonth -= 1;
    }
  }

  @Listen('b2b-date-picker-next-month')
  nextMonth(): void {
    if (this.viewedMonth === 11) {
      this.viewedMonth = 0;
      this.viewedYear += 1;
    } else {
      this.viewedMonth += 1;
    }
  }

  @Listen('b2b-date-picker-month-selected')
  handleMonthSelected(event: CustomEvent<MonthSelectedEventDetail>): void {
    this.viewedMonth = event.detail.value + 1;
    this.viewMode = DatePickerView.Days;
  }

  @Listen('b2b-date-picker-year-selected')
  handleYearSelected(event: CustomEvent<YearSelectedEventDetail>): void {
    this.viewedYear = event.detail.value;
    this.viewMode = DatePickerView.Days;
  }

  @Listen('b2b-date-picker-escape')
  handleEscapePress(): void {
    this.toggleOpen(false);
  }

  @Listen('b2b-date-selected')
  handleDateSelected(event: CustomEvent<DatePickerEventDetail>): void {
    event.preventDefault();
    this.selectDate(event.detail.selectedDate);
  }

  private selectDate(selected: Dateish): void {
    if (this.tempDate == null) {
      this.tempDate = splitDate(selected);
    } else if (dateAsNumber(selected) >= dateAsNumber(this.tempDate)) {
      this.setDate([toDate(this.tempDate), toDate(selected)]);
    } else {
      this.setDate([toDate(selected), toDate(this.tempDate)]);
    }
  }

  private setDate(selected: DateRange | undefined): void {
    this.internalDateRange = selected;
    this.toggleOpen(false);
    if (selected == null) {
      this.b2bClear.emit();
    } else {
      this.b2bSelected.emit({ selected });
    }
  }

  private toggleOpen(open: boolean = !this.open): void {
    this.open = open;
    if (!open) {
      this.tempDate = undefined;
      this.resetViewedDate();
    }
  }

  private resetViewedDate(): void {
    const startDate = this.internalDateRange?.[0];
    if (startDate == null) {
      const now = new Date();
      this.viewedYear = now.getFullYear();
      this.viewedMonth = now.getMonth() + 1;
      this.viewedDay = now.getDate();
    } else {
      const [year, month, day] = splitDate(startDate);
      this.viewedYear = year;
      this.viewedMonth = month;
      this.viewedDay = day;
    }
    this.viewMode = DatePickerView.Days;
  }

  // #endregion
  // #region Component-Hooks

  componentWillLoad() {
    this.internalDateRange = toDateRange(this.dateRange);
    this.resetViewedDate();
  }

  render() {
    return (
      <Host>
        <div class="b2b-date-range-picker">
          <b2b-input-label required={this.required}>
            {this.label}
          </b2b-input-label>
          <div
            class={{
              'b2b-date-range-picker-input-wrapper': true,
              'b2b-date-range-picker-input-wrapper--error': this.invalid,
            }}
            onClick={() => this.toggleOpen()}>
            <b2b-paragraph margin={false}>
              {toDateRangeString(this.internalDateRange, '')}
            </b2b-paragraph>
            <div class="b2b-icons">
              {this.internalDateRange && (
                <div
                  tabIndex={0}
                  onClick={() => this.setDate(undefined)}
                  onKeyDown={(event: KeyboardEvent) => {
                    if (event.key === 'Enter') {
                      this.setDate(undefined);
                    }
                  }}>
                  <b2b-icon-100
                    icon="b2b_icon-close"
                    aria-label="clear input"
                    clickable={true}
                  />
                </div>
              )}

              <div
                tabindex={0}
                onClick={() => this.toggleOpen()}
                onKeyDown={event => {
                  if (event.key === 'Enter') {
                    this.toggleOpen();
                  }
                }}
                class="b2b-event-icon">
                <b2b-icon-100
                  aria-label={
                    this.open ? 'close date picker' : 'open date picker'
                  }
                  icon="b2b_icon-event"
                  clickable={true}
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class={{
            'b2b-date-range-picker-body': true,
            'b2b-date-range-picker-body--hidden': !this.open,
          }}>
          <div class="b2b-date-range-picker-body--calendar">
            {this.viewMode === DatePickerView.Days && (
              <div>
                <b2b-date-picker-header
                  language={this.language}
                  selectedMonth={this.viewedMonth - 1}
                  selectedYear={this.viewedYear}
                />
                <b2b-date-picker-days-header language={this.language} />
                <b2b-date-range-picker-days
                  viewedDate={
                    [this.viewedYear, this.viewedMonth, this.viewedDay] as const
                  }
                  firstDate={
                    this.highlightDateRange?.[0] ??
                    this.tempDate ??
                    this.internalDateRange?.[0]
                  }
                  secondDate={
                    this.highlightDateRange?.[1] ??
                    (this.tempDate == null
                      ? this.internalDateRange?.[1]
                      : undefined)
                  }
                  disableDates={day =>
                    this.disableDates(
                      toDate([this.viewedYear, this.viewedMonth, day]),
                    )
                  }
                  preview={this.highlightDateRange != null}
                />
              </div>
            )}
            {this.viewMode === DatePickerView.Months && (
              <b2b-date-picker-months
                language={this.language}
                selectedMonth={this.viewedMonth - 1}
              />
            )}
            {this.viewMode === DatePickerView.Years && (
              <b2b-date-picker-years selectedYear={this.viewedYear} />
            )}
          </div>
          {Object.keys(this.presets).length > 0 && (
            <div class="b2b-date-range-picker-body--presets">
              {Object.entries(this.presets).map(([text, value]) => (
                <b2b-paragraph
                  class={{
                    'b2b-date-range-picker-body--presets--selected':
                      isDateRangeEqual(this.internalDateRange, value),
                  }}
                  margin={false}
                  onClick={() => this.setDate(value)}
                  onMouseOver={() => (this.highlightDateRange = value)}
                  onMouseOut={() => (this.highlightDateRange = undefined)}>
                  {text}
                </b2b-paragraph>
              ))}
            </div>
          )}
        </div>
        {this.open && (
          <div
            class="b2b-date-range-picker--backdrop"
            onClick={() => this.toggleOpen(false)}></div>
        )}
        {(this.error != null || this.hint != null) && (
          <span
            class={{
              'b2b-date-range-picker-hint': true,
              'b2b-date-range-picker-hint--error': this.invalid,
            }}>
            {this.invalid ? this.error : this.hint}
          </span>
        )}
      </Host>
    );
  }

  // #endregion
}
