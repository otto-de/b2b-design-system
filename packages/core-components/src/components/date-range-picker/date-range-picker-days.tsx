import type { ComponentInterface, EventEmitter } from '@stencil/core';
import {
  Component,
  Element,
  Event,
  h,
  Host,
  Listen,
  Prop,
  State,
} from '@stencil/core';
import type {
  DateSelectedEventDetail,
  EscapePressed,
} from '../../utils/interfaces/form.interface';
import { dateAsNumber, Dateish, splitDate } from './date-range-picker-utils';

const keys = {
  ARROW_UP: 'ArrowUp',
  ARROW_DOWN: 'ArrowDown',
  ARROW_RIGHT: 'ArrowRight',
  ARROW_LEFT: 'ArrowLeft',
  ESC: 'Escape',
  ENTER: 'Enter',
};

@Component({
  tag: 'b2b-date-range-picker-days',
  styleUrl: 'date-range-picker-days.scss',
  shadow: true,
})
export class B2bDatePickerDays implements ComponentInterface {
  // #region Properties
  /**
   * The root of this element.
   */
  @Element() host: HTMLB2bDateRangePickerDaysElement;

  /**
   * The currently shown date (year+month).
   */
  @Prop() viewedDate: Dateish = new Date();

  /**
   * The first date of the selection.
   * This may be absent if no dates have been selected yet.
   * This is always smaller/earlier than the second date.
   */
  @Prop() firstDate?: Dateish;
  /**
   * The second date of the selection.
   * This may be absent if none or only one date has been selected yet.
   * This is always bigger/later than the first date.
   */
  @Prop() secondDate?: Dateish;

  /**
   * Function used to determine, whether a day should be disabled e.g. because it is in the past.
   * This is only used for the start and end dates, it is valid to have a disabled day in the range of start and end date.
   *
   * @param day The day of the month to check (1-indexed).
   * @returns True, if the day should be disabled. False otherwise.
   *
   * @default (day) => false
   */
  @Prop() disableDates: (day: number) => boolean = () => false;

  /**
   * Whether the shown values are a preview and not the actual values (force-hover state).
   *
   * @default false
   */
  @Prop() preview: boolean = false;

  // #endregion
  // #region Events

  /**
   * Event emitted on escape press.
   */
  @Event({ eventName: 'b2b-date-picker-escape' })
  private b2bDatePickerEscape: EventEmitter<EscapePressed>;

  /**
   * Event emitted on selecting date.
   */
  @Event({ eventName: 'b2b-date-selected' })
  private b2bDateSelected: EventEmitter<DateSelectedEventDetail>;

  // #endregion
  // #region State

  @State() viewedDay: number = 0;

  // #endregion
  // #region Logic

  @Listen('keydown')
  handleKeyDown(event: KeyboardEvent) {
    event.preventDefault();
    const [year, month] = splitDate(this.viewedDate);
    let day = this.viewedDay;
    switch (event.key) {
      case keys.ARROW_LEFT:
        day -= 1;
        break;
      case keys.ARROW_RIGHT:
        day += 1;
        break;
      case keys.ARROW_UP:
        day -= 7;
        break;
      case keys.ARROW_DOWN:
        day += 7;
        break;
      case keys.ENTER:
        if (!this.disableDates(day)) {
          this.b2bDateSelected.emit({
            selectedDate: new Date(year, month - 1, day),
          });
        }
        return;
      case keys.ESC:
        this.b2bDatePickerEscape.emit();
        return;
      default:
        return;
    }

    const dates = this.getAllDates();
    if (day < 1) {
      day = dates.length;
    }

    if (day > dates.length) {
      day = 1;
    }

    if (this.disableDates(day)) {
      return;
    }

    this.viewedDay = day;

    this.focusCurrentDate(dates[day - 1]);
  }

  private getAllDates = (): HTMLDivElement[] => {
    return Array.from(
      this.host.shadowRoot.querySelectorAll('.b2b-date-range-picker-day'),
    ) as HTMLDivElement[];
  };

  private focusCurrentDate = (date: HTMLDivElement) => {
    const dates = this.getAllDates();
    dates.forEach(element => {
      if (date.className.includes('b2b-date-range-picker-day--disabled')) {
        return;
      }
      element.setAttribute('tabindex', element === date ? '0' : '-1');
    });
    if (!date.className.includes('b2b-date-range-picker-day--disabled')) {
      date.focus();
    }
  };

  private renderDatePickerDays = () => {
    const [year, month] = splitDate(this.viewedDate);
    const day = this.viewedDay;
    const currentNumber = dateAsNumber([year, month, day]);
    const todayNumber = dateAsNumber(new Date());
    const startNumber = dateAsNumber(this.firstDate);
    const endNumber = dateAsNumber(this.secondDate);
    let totalDaysInMonth = new Date(
      year,
      month - 1 + 1, // Next month
      0, // The day before
    ).getDate();

    let days = [];

    /** Populate the days before the first day of the month with empty divs. */
    let dayOfWeekOffset = (new Date(year, month - 1, 1).getDay() + 6) % 7; // +6 to make Monday the first day of the week
    for (let i = 0; i < dayOfWeekOffset; i++) {
      days.push(<div></div>);
    }

    /** Populate days array with day numbers. */
    for (let i = 1; i <= totalDaysInMonth; i++) {
      let givenDate = [year, month, i] as const;
      const givenDateNumber = dateAsNumber(givenDate);
      const disabled = this.disableDates(i);
      days.push(
        <div
          class={{
            'b2b-date-range-picker-day': true,
            'b2b-date-range-picker-day--disabled': disabled,
            'b2b-date-range-picker-day--today': givenDateNumber === todayNumber,
            'b2b-date-range-picker-day--selected':
              givenDateNumber === startNumber || givenDateNumber === endNumber,
            'b2b-date-range-picker-day--preview':
              endNumber == null || this.preview,
            'b2b-date-range-picker-day--range':
              startNumber != null &&
              ((startNumber < givenDateNumber &&
                givenDateNumber < (endNumber ?? currentNumber)) ||
                (startNumber > givenDateNumber &&
                  givenDateNumber > (endNumber ?? currentNumber))),
          }}
          onClick={() => {
            if (!disabled) {
              this.b2bDateSelected.emit({
                selectedDate: new Date(year, month - 1, i),
              });
            }
          }}
          onMouseEnter={(event: MouseEvent) => {
            this.viewedDay = i;
            (event.target as HTMLElement).focus();
            event.preventDefault();
          }}
          tabindex={givenDateNumber === currentNumber ? 0 : -1}
          role="gridcell"
          aria-label={`Date ${i} ${month} ${year}`}>
          {i}
        </div>,
      );
    }

    return days;
  };

  private setFocusOnSelectedDayOrFirstDay() {
    const dates = this.getAllDates();
    if (dates.length > 0) {
      const [currentYear, currentMonth] = splitDate(this.viewedDate);
      const [selectedYear, selectedMonth, selectedDay] = splitDate(
        this.firstDate ?? new Date(),
      );

      if (
        currentYear == selectedYear &&
        currentMonth == selectedMonth &&
        this.viewedDay === selectedDay
      ) {
        this.focusCurrentDate(dates[this.viewedDay - 1]);
      } else {
        this.focusCurrentDate(dates[0]);
      }
    }
  }

  // #endregion
  // #region Component-Hooks

  componentWillLoad(): void {
    let [, , day] = splitDate(this.viewedDate);
    this.viewedDay = day;
  }

  componentDidLoad(): void {
    this.setFocusOnSelectedDayOrFirstDay();
  }

  render() {
    return (
      <Host>
        <div class="b2b-date-range-picker-days">
          {this.renderDatePickerDays()}
        </div>
      </Host>
    );
  }

  // #endregion
}
