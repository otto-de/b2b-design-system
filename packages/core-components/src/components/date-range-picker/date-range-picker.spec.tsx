import { newSpecPage } from '@stencil/core/testing';
import { h } from '@stencil/core';
import { B2bDateRangePicker } from './date-range-picker';
import { ParagraphComponent } from '../paragraph/paragraph';

it('should render the date-range-picker', async () => {
  const page = await newSpecPage({
    components: [B2bDateRangePicker],
    template: () => <b2b-date-range-picker></b2b-date-range-picker>,
  });
  expect(page.root).toMatchSnapshot();
});

it('should display custom properties', async () => {
  const page = await newSpecPage({
    components: [B2bDateRangePicker, ParagraphComponent],
    template: () => (
      <b2b-date-range-picker
        label="Test"
        hint="Test"
        error="Error"></b2b-date-range-picker>
    ),
  });
  expect(page.root).toMatchSnapshot();
});
