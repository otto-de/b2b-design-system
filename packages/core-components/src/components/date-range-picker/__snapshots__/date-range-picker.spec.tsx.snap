// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display custom properties 1`] = `
<b2b-date-range-picker>
  <mock:shadow-root>
    <div class="b2b-date-range-picker">
      <b2b-input-label>
        Test
      </b2b-input-label>
      <div class="b2b-date-range-picker-input-wrapper">
        <b2b-paragraph>
          <mock:shadow-root>
            <p class="b2b-paragraph b2b-paragraph--align-left b2b-paragraph--no-margin b2b-paragraph--size-100 b2b-paragraph--variant-black-100 b2b-paragraph--weight-normal" style="display: block;">
              <slot></slot>
            </p>
          </mock:shadow-root>
        </b2b-paragraph>
        <div class="b2b-icons">
          <div class="b2b-event-icon" tabindex="0">
            <b2b-icon-100 aria-label="open date picker" clickable="" icon="b2b_icon-event"></b2b-icon-100>
          </div>
        </div>
      </div>
    </div>
    <div class="b2b-date-range-picker-body b2b-date-range-picker-body--hidden">
      <div class="b2b-date-range-picker-body--calendar">
        <div>
          <b2b-date-picker-header language="de" selectedmonth="9" selectedyear="2025"></b2b-date-picker-header>
          <b2b-date-picker-days-header language="de"></b2b-date-picker-days-header>
          <b2b-date-range-picker-days></b2b-date-range-picker-days>
        </div>
      </div>
    </div>
    <span class="b2b-date-range-picker-hint">
      Test
    </span>
  </mock:shadow-root>
</b2b-date-range-picker>
`;

exports[`should render the date-range-picker 1`] = `
<b2b-date-range-picker>
  <mock:shadow-root>
    <div class="b2b-date-range-picker">
      <b2b-input-label>
        Zeitraum ausw√§hlen
      </b2b-input-label>
      <div class="b2b-date-range-picker-input-wrapper">
        <b2b-paragraph></b2b-paragraph>
        <div class="b2b-icons">
          <div class="b2b-event-icon" tabindex="0">
            <b2b-icon-100 aria-label="open date picker" clickable="" icon="b2b_icon-event"></b2b-icon-100>
          </div>
        </div>
      </div>
    </div>
    <div class="b2b-date-range-picker-body b2b-date-range-picker-body--hidden">
      <div class="b2b-date-range-picker-body--calendar">
        <div>
          <b2b-date-picker-header language="de" selectedmonth="9" selectedyear="2025"></b2b-date-picker-header>
          <b2b-date-picker-days-header language="de"></b2b-date-picker-days-header>
          <b2b-date-range-picker-days></b2b-date-range-picker-days>
        </div>
      </div>
    </div>
  </mock:shadow-root>
</b2b-date-range-picker>
`;
